### PR TITLE
Adding schema validation for template configuration

### PR DESCRIPTION
**Pull request by SpyzzVVarun**
Sorry for the unnecessary amount of commits you see.

The main changes are as follows:
1.  Added jsonschema requirement to requirements.txt
2. Added code to template.py that updates imports, open_template_with_defaults function
3. Added code to core.py that updates process_dir function
4. Added schema.json
5. Added code to file.py that updates imports, adds load_schema and validate_json functions
6. Added a sampleError folder which contains a template.json file with some errors

P.S. I am very new to Open Source Contribution and am open to any and all feedback.

### CODE REVIEW

File: src/defaults/schema.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+         "type":"integer"
+       },
+       "minItems": 2,
+       "maxItems": 2
+     },
+    "preProcessors": {
+       "description":"Custom configuration values to use in the template's directory"
+     },
+  "concatenations": {
+       "description":"The Contatenations parameter is a way to tell OMRChecker which fields need to be joined together before outputting into the csv"
```
Comment (Udayraj123) [inline src/defaults/schema.json:30]: typo

File: src/defaults/schema.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "https://example.com/product.schema.json",
+  "title": "Input",
+  "description": "OMRchecker input image schema",
```
Comment (Udayraj123) [inline src/defaults/schema.json:5]: OMRChecker

File: src/core.py [hunk 3]
```diff
@@ -55,8 +55,9 @@ def process_dir(root_dir, curr_dir, args, template=None):
 
     # Update local template (in current recursion stack)
     local_template_path = curr_dir.joinpath(constants.TEMPLATE_FILENAME)
+    local_schema_path = constants.SCHEMA_DEFAULTS_PATH
     if os.path.exists(local_template_path):
-        template = Template(local_template_path, PROCESSOR_MANAGER.processors)
+        template = Template(local_schema_path,local_template_path, PROCESSOR_MANAGER.processors)
```
Comment (Udayraj123) [inline src/core.py:60]: spacing after ','

File: src/template.py [hunk 4]
```diff
@@ -8,17 +8,23 @@
...
 
 
-def open_template_with_defaults(template_path):
+def open_template_with_defaults(schema_path,template_path):
     user_template = load_json(template_path)
-    return OVERRIDE_MERGER.merge(TEMPLATE_DEFAULTS, user_template)
-
+    user_template = OVERRIDE_MERGER.merge(TEMPLATE_DEFAULTS, user_template)
+    is_valid, msg = validate_json(user_template,schema_path)
+    if is_valid:
```
Comment (Udayraj123) [inline src/template.py:23]: add new line above

File: samples/sampleError/template.json [hunk 5]
```diff
@@ -0,0 +1,274 @@
+{
```
Comment (Udayraj123) [inline samples/sampleError/template.json:1]: remove `sampleError` folder. Will cover in separte pytest PR

**Review by Udayraj123 — APPROVED** [review]

**Review by Udayraj123 — CHANGES_REQUESTED** [review]
resolve above comments

**Comment by Udayraj123** [comment]
Need a separate issue for enhancements in schema:
1. All labels in all of the QBlocks should be covered in either Singles or concatenations. 
2. Also none of the labels in qBlocks should overlap (show warning in this case)
3. User friendly messages: suggestion to convert key into camelCase (in which case validation error should mention that).

**Comment by SpyzzVVarun** [comment]
Error Message for sampleError:

error_message

**Comment by Udayraj123** [comment]
use table for outputting errors: 
example https://pythonfusion.com/table-on-console-python/

File: src/defaults/schema.json (hunk 1 above)
Comment (SpyzzVVarun) [reply src/defaults/schema.json]: Done

File: src/defaults/schema.json (hunk 2 above)
Comment (SpyzzVVarun) [reply src/defaults/schema.json]: Done

File: src/template.py (hunk 4 above)
Comment (SpyzzVVarun) [reply src/template.py:24]: Added newline and other spacings in template.py

File: src/core.py (hunk 3 above)
Comment (SpyzzVVarun) [reply src/core.py]: Done

**Comment by SpyzzVVarun** [comment]
> use table for outputting errors: example https://pythonfusion.com/table-on-console-python/

Is this part of this PR only or included in the enhancement issue?

File: samples/sampleError/template.json (hunk 5 above)
Comment (SpyzzVVarun) [reply samples/sampleError/template.json]: Deleted

File: src/core.py [hunk 6]
```diff
@@ -55,8 +55,9 @@ def process_dir(root_dir, curr_dir, args, template=None):
 
     # Update local template (in current recursion stack)
     local_template_path = curr_dir.joinpath(constants.TEMPLATE_FILENAME)
+    local_schema_path = constants.SCHEMA_DEFAULTS_PATH
     if os.path.exists(local_template_path):
-        template = Template(local_template_path, PROCESSOR_MANAGER.processors)
+        template = Template(local_schema_path, local_template_path, PROCESSOR_MANAGER.processors)
```
Comment (Udayraj123) [inline src/core.py:60]: Directly use `constants.SCHEMA_DEFAULTS_PATH`

File: src/utils/file.py [hunk 7]
```diff
@@ -1,7 +1,43 @@
...
+          validator = error.validator
+          msg = error.message
+
+          if validator == 'required':
+            table.add_row(re.findall(r"'(.*?)'",msg)[0], msg + ". Make sure the spelling of the key is correct and it is in camelCase")
+          else:
+            table.add_row(key[0], msg)
+        console = Console()
+        console.print(table)
+        err = "Given JSON data is Invalid"
```
Comment (Udayraj123) [inline src/utils/file.py:38]: Provided Template JSON is Invalid

File: src/utils/file.py [hunk 8]
```diff
@@ -1,7 +1,43 @@
...
+        return False, err
+
+    message = "Given JSON data is Valid"
```
Comment (Udayraj123) [inline src/utils/file.py:41]: Template JSON validated successfully...

**Review by Udayraj123 — CHANGES_REQUESTED** [review]

File: src/utils/file.py [hunk 9]
```diff
@@ -1,7 +1,43 @@
...
+        table = Table(show_lines = True)
+        table.add_column("Key", style="cyan", no_wrap=True)
+        table.add_column("Error", style="magenta")
+
+        v = Draft202012Validator(execute_api_schema)
+        errors = sorted(v.iter_errors(json_data), key=lambda e: e.path)
+        for error in errors:
+          key = error.path
+          validator = error.validator
+          msg = error.message
```
[the remark this answers is not in the document]
Comment (Udayraj123) [reply src/utils/file.py:30]: Or use this https://stackoverflow.com/a/63600600/6242649

File: src/defaults/schema.json [hunk 10]
```diff
@@ -0,0 +1,41 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "https://example.com/product.schema.json",
```
[the remark this answers is not in the document]
Comment (SpyzzVVarun) [reply src/defaults/schema.json]: Done

File: src/defaults/schema.json [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  "title": "Input",
```
[the remark this answers is not in the document]
Comment (SpyzzVVarun) [reply src/defaults/schema.json]: Done

File: src/defaults/schema.json [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+  },
+  "singles": {
+       "description":"The remaining fields(in order) whose readings shall be forwarded directly in the output csv",
+       "type": "array",
+       "items": {
+         "type":"string"
+       }
+     }
+  },
+  "required":["dimensions","bubbleDimensions","concatenations","singles","preProcessors"]
```
[the remark this answers is not in the document]
Comment (SpyzzVVarun) [reply src/defaults/schema.json]: Done

File: src/utils/file.py (hunk 8 above)
Comment (SpyzzVVarun) [reply src/utils/file.py:41]: Done

File: src/utils/file.py (hunk 7 above)
Comment (SpyzzVVarun) [reply src/utils/file.py:38]: Done

File: src/template.py [hunk 13]
```diff
@@ -8,17 +8,24 @@
...
-    return OVERRIDE_MERGER.merge(TEMPLATE_DEFAULTS, user_template)
-
+    user_template = OVERRIDE_MERGER.merge(TEMPLATE_DEFAULTS, user_template)
+    is_valid, msg = validate_json(user_template, schema_path)
+    
+    if is_valid:
+      return user_template
+    else:
+      print(msg)
+      sys.exit(1)
```
[the remark this answers is not in the document]
Comment (SpyzzVVarun) [reply src/template.py:28]: Done

File: src/utils/file.py [hunk 14]
```diff
@@ -1,7 +1,43 @@
...
+    try:
+        validate(instance=json_data, schema=execute_api_schema)
+    except jsonschema.exceptions.ValidationError as err:
+        print("\nValidating template.json ...")
+
+        table = Table(show_lines = True)
+        table.add_column("Key", style="cyan", no_wrap=True)
+        table.add_column("Error", style="magenta")
+
+        v = Draft202012Validator(execute_api_schema)
```
[the remark this answers is not in the document]
Comment (SpyzzVVarun) [reply src/utils/file.py]: Done

File: src/core.py (hunk 6 above)
Comment (SpyzzVVarun) [reply src/core.py]: Done

File: src/defaults/schema.json [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    "bubbleDimensions": {
+      "description": "The dimensions of the overlay bubble area",
+      "type": "array",
+      "items": {
+        "type": "integer"
+      },
+      "minItems": 2,
+      "maxItems": 2
+    },
+    "preProcessors": {
```
Comment (Udayraj123) [inline src/defaults/schema.json:26]: array of objects

File: src/utils/file.py [hunk 16]
```diff
@@ -1,7 +1,42 @@
 import json
-
+import jsonschema
+from jsonschema import validate, Draft202012Validator
+from rich.console import Console
+from rich.table import Table
+import re
+from ..constants import SCHEMA_DEFAULTS_PATH
```
Comment (Udayraj123) [inline src/utils/file.py:7]: use absolute path, `..` is a bit error-prone

**Review by Udayraj123 — CHANGES_REQUESTED** [review]

File: src/defaults/schema.json (hunk 15 above)
Comment (SpyzzVVarun) [reply src/defaults/schema.json:26]: Done

**Review by Udayraj123 — APPROVED** [review]

**Comment by Udayraj123** [comment]
Fixes https://github.com/Udayraj123/OMRChecker/issues/43

**Pull request merged**